toolchain/lower/file_context.h

```diff
@@ -165,6 +184,69 @@ class FileContext {
 
   auto BuildVtable(const SemIR::Class& class_info) -> llvm::GlobalVariable*;
 
+  // Track which specifics were lowered for each generic. These are added one
+  // by one while lowering their definitions.
+  auto AddLoweredSpecificForGeneric(SemIR::GenericId generic_id,
+                                    SemIR::SpecificId specific_id) {
+    lowered_specifics_[generic_id.index].push_back(specific_id);
+  }
+
+  // Initialize a return a SpecificFunctionFingerprint* instance for specific.
+  // The internal of the fingerprint are populated during and after lowering
+  // the function body of that specific.
+  auto InitializeFingerprintForSpecific(SemIR::SpecificId specific_id)
+      -> SpecificFunctionFingerprint* {
+    if (!specific_id.has_value()) {
+      return nullptr;
+    }
+    return &lowered_specific_fingerprint_[specific_id.index];
+  }
+
+  // Entry point for coalescing equivalent specifics. Two function definitions,
+  // from the same generic, with different specific_ids are considered
+  // equivalent if, at the LLVM level, one can be replaced with the other, with
+  // no change in behavior. All LLVM types and instructions must be equivalent.
+  auto CoalesceEquivalentSpecifics() -> void;
+
+  // While coalescing specifics, compare the function types for two specifics.
+  // This uses a fingerprint generated for each function type.
+  auto AreFunctionTypesEquivalent(SemIR::SpecificId specific_id1,
+                                  SemIR::SpecificId specific_id2) -> bool;
+
+  // While coalescing specifics, compare the function bodies for two specifics.
+  // This uses fingerprints generated during lowering of the function body.
+  auto AreFunctionBodiesEquivalent(
+      SemIR::SpecificId specific_id1, SemIR::SpecificId specific_id2,
+      Set<std::pair<SemIR::SpecificId, SemIR::SpecificId>>&
+          visited_equivalent_specifics) -> bool;
+
+  // When an equivalence is found, update the canonical specific to use for
```

Comment:
This is being handled an equivalence, it's not checking for it, right? Maybe phrase as "Given an equivalent pair of specifics" (emphasizing "pair" because of your parameter name), instead of "When an equivalence is found".

Otherwise, what happens when an equivalence isn't found?

```suggestion
  // Given an equivalent pair of specifics, update the canonical specific to use for
```